[doc] How To Request Support document stab (#9288)
huggingface/transformers: ISSUES.md

   The "who to tag for what domain" part of the Issue template is there to help users direct their questions to the right developers who are designated maintainers of project's specific domains. They can then decide at their own discretion to tag other developers if they feel it'd help move the Issue forward.
   
   We currently don't have a triage service and we trust your capacity to identify the right domain and thus the persons to tag in your Issue. If you are not sure, please use the forums to ask for guidance.
   
   When in doubt, err on the side of not tagging a given person. If you tag multiple people out of context or permission don't be surprised if you get no response at all. Please remember that every time you tag someone, they get a notification and you're taking their time without their permission. Please be sensitive to that.

   If you got helped by one of the developers in the past please don't tag them in future Issues, unless they are listed in the Issue template for the domain you are asking about or that developer gave you an explicit permission to tag them in future Issues.
   

>    If you see a certain developer doing multiple and/or recent commits into a specific area of the project that you feel is relevant to your Issue, it is not a good reason to tag them. Various developers may be fixing things that prevent them from moving forward, but often their work is focused on a totally different domain. And while they may or may not know how to help you with the problem at hand, it would benefit the whole community much more if they focus on the domain of their unique expertise.

Reviewer: I somewhat disagree with that statement. If the issue respects the rest of this document, then it is well structured, the problem is well explained, and a member of the team that recently worked on that area of the project will most likely be the best suited to answer the issue. It's a very fast moving project so members that worked on an aspect of the project one year ago are not experts on that area anymore, but users that recently edited/worked on that area most likely have the best understanding of that part.

Author: But that's why the Issue/PR templates have designated domain "owners" - those are the people who know the domain - their role could be to dispatch the issue to the person they know is most familiar with the issue at hand. Asking users to figure out who to tag is just unfair to the user and is asking too much, IMHO. 

I like how some projects have a triage process - e.g. pytorch - the user is not requested to tag anybody. Whoever is doing the triage - and they switch labels to handle this process - tags one or more developers who they know would be the most relevant developer to take care of the issue at hand. Surely, pytorch is much bigger and has many more issues and developers to handle such process, but what I like is that the project takes care of this on behalf of users. 

So perhaps the most user-friendly solution is just that - don't tag - we will tag for you. 

This is just a suggestion as everything else I have written in this document. Please feel free to edit/remove/rewrite as you feel fit.

somewhat related - if user-tags-developers is a thing - then using labels would make it easier to find things for both developers and users. So users could hit a label of their interest and see which developer does most work in this area as of recent.

Reviewer: I wasn't thinking that we had to specify "You should look who last edited these files", but rather that we don't need to put in the document "Do not look at who last edited these files to tag them". In any case, I agree with your points. Tagging the domain owner so that he can tag others if they are more knowledgeable than the owner is good enough for me.

## The Forums

[The user forums](https://discuss.huggingface.co/) are supported by the wide community of the library users and backed up by developers when needed.

If you have a difficulty with deploying this library or some questions, or you'd like to discuss a new feature, please first consider discussing those things at the forums. Only when you feel your subject matter has been crystalized and you still need support from the library developers do proceed to file an [Issue](https://github.com/huggingface/transformers/issues).

In particular all "Please explain" questions or objectively very user-specific feature requests belong to the forums. Here are some example of such questions:

* "I would like to use a BertModel within a RL-Agent for a customer support service. How can I use a BertForMaskedLM in my ChatBotModel?"

* "Could you please explain why T5 has no positional embedding matrix under T5Model?"

* "How should I set my generation parameters for translation?"

* "How to train T5 on De->En translation?"

## The github Issues

> Everything which hints at a bug should be opened as an [Issue](https://github.com/huggingface/transformers/issues).

Author: Well, I did this on purpose to make it clear that it's referring to the github Issue. 

I can easily lower-case those. Yes?

> Everything which hints at a bug should be opened as an [Issue](https://github.com/huggingface/transformers/issues).

Author: done

You are not required to read the following guidelines before opening an Issue. However, if you notice that your issue is not being addressed, chances are that the developers have one or several difficulties with the quality of that Issue. In which case please read the following guidelines and adjust your Issue accordingly.

The guidelines:

1. Before posting an Issue, first search for already posted issues, since chances are someone has already asked a similar question before you. 

    If you use Google your search query should be:
    
    ```
    "huggingface" "transformers" your query 
    ```
    

>     The first two quoted words tell Google to limit the search to the context of the Huggingface Transformers. The remainder is your query - most commonly this would be an assert line if the software fails with an assert. 

Reviewer 2: Not sure assert line is going to help beginners a lot.
```suggestion
    The first two quoted words tell Google to limit the search to the context of the Huggingface Transformers. The remainder is your query - most commonly this would be a simplified version of your error message.
```

Author: I hear you. But I'm not sure how "a simplified version of your error message" is any more clear.

Perhaps leaving the original and using an example to show what is meant by an assert and how to convert it into a useful search query?

Author: e.g.:
```
If you have an error message like:
  ```python
   $ python -c 'import transformers'
   Traceback (most recent call last):
     File "<string>", line 1, in <module>
     File "/transformers/src/transformers/__init__.py", line 34, in <module>
       from . import dependency_versions_check
     File "/transformers/src/transformers/dependency_versions_check.py", line 34, in <module>
       from .file_utils import is_tokenizers_available
     File "/transformers/src/transformers/file_utils.py", line 40, in <module>
       from tqdm.auto import tqdm
   ModuleNotFoundError: No module named 'tqdm.auto'
   ```
   
   Search first for:
   1. search first for "huggingface" "transformers" "ModuleNotFoundError: No module named 'tqdm.auto'"
   2. if you don't find relevant results, then search for just "ModuleNotFoundError: No module named 'tqdm.auto'"
   3. and finally if nothing comes up remove the quotes

If the error includes any error messages that include bits of your filesystem, always remove those in the search query since other users will not have the same bits as yours. In such cases always remove double quotes since your query is not longer exact. For example:

python -c 'open("/tmp/wrong_path.txt", "r")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/wrong_path.txt'

Here you'd search for just: `FileNotFoundError: [Errno 2] No such file or directory`

```

Reviewer 2: Yes showing an example is great!

Author: Done. I also rewrote the bit on assert line and explained what it means.

>     The results of such query will typically match github Issues, huggingface forums and StackExchange and people's blogs.

Reviewer 2: ```suggestion
    The results of such a query will typically match GitHub issues, Hugging Face forums, StackExchange, and other user's blogs.
```

    
    If you find relevant hints, you may choose to continue the discussion there if you have follow up questions. 
    
    If what you found is similar but doesn't quite answer your problem, please, post a new Issue and do include links to similar Issues or forum discussions you may have found.
    

> 2. Keep the Issue short, providing the information that you think will aid the developers to understand your situation. Put yourself in the shoes of the person who has never seen your code or knows anything about your custom setup. This mental exercise will help a lot for you to know what to share and what not.

Reviewer 2: ```suggestion
2. Keep the Issue short, providing the information that you think will aid the developers to understand your situation. Put yourself in the shoes of the person who has never seen your code or knows anything about your custom setup. This mental exercise will help a lot for you to guess what/what not to share.
```

Author: why is it bad for users to know? this is the point of such guidelines - so that they can know and not guess.

do you feel that we put too much responsibility for when we ask them to know something?

how about "This mental exercise will help to develop an intuition to what/what not to share"?

Reviewer 2: It's just to use another word since `knows` is used in the sentence before.

Author: sorry, I don't see the connection - they are 2 different contexts, 2 totally different "know"

Author: I went with intuition, as it's even better I think.

3. If there is a software failure, always provide the full traceback, for example:

   ```python
   $ python -c 'import transformers'
   Traceback (most recent call last):
     File "<string>", line 1, in <module>
     File "/transformers/src/transformers/__init__.py", line 34, in <module>
       from . import dependency_versions_check
     File "/transformers/src/transformers/dependency_versions_check.py", line 34, in <module>
       from .file_utils import is_tokenizers_available
     File "/transformers/src/transformers/file_utils.py", line 40, in <module>
       from tqdm.auto import tqdm
   ModuleNotFoundError: No module named 'tqdm.auto'
   ```
   

>    As compared to providing just the assert message, e.g.:

Reviewer 2: ```suggestion
   As compared to providing just the last line of the error message, e.g.:
```

9. Please do not ask us to reproduce an Issue with your custom data, since we don't have it. So, either you should use some existing dataset supported by HF datasets or you need to supply a code that generates a small sample on the fly, or some another quick and simple way to get it.

   Please do not send us any non-public domain data that may require a license or a permission to be used.

10. Do not tag multiple developers on the Issue unless you know this is expected, either because you asked them and they gave you an explicit permission to tag them or the Issue template instructs you to do so. 

   The "who to tag for what domain" part of the Issue template is there to help users direct their questions to the right developers who are designated maintainers of project's specific domains. They can then decide at their own discretion to tag other developers if they feel it'd help move the Issue forward.
   
   We currently don't have a triage service and we trust your capacity to identify the right domain and thus the persons to tag in your Issue. If you are not sure, please use the forums to ask for guidance.
   

>    If you tag multiple people out of context or permission don't be surprised if you get no response at all. Please remember that every time you tag someone, they get a notification and you're taking their time without their permission. Please be sensitive to that.

Reviewer 2: ```suggestion
   When in doubt, err on the side of not tagging a given person. If you tag multiple people out of context or permission don't be surprised if you get no response at all. Please remember that every time you tag someone, they get a notification and you're taking their time without their permission. Please be sensitive to that.
```

    

12. If you are cross-referencing a specific comment in a given thread or another Issue, always link to that specific comment, rather than using the Issue link. If you do the latter it could be quite impossible to find which specific comment you're referring to.

    To get the link to the specific comment do not copy the url from the location bar of your browser, but instead, click the `...` icon in the upper right corner of the comment and then select "Copy Link".
    
    For example the first link is a link to an Issue, and the second to a specific comment in the same Issue:
    
    1. https://github.com/huggingface/transformers/Issues/9257
    2. https://github.com/huggingface/transformers/Issues/9257#Issuecomment-749945162

    
13. If you are replying to a last comment, it's totally fine to make your reply with just your comment in it. The readers can follow the information flow here.

    But if you're replying to a comment that happened some comments back it's always a good practice to quote just the relevant lines you're replying it. The `>` is used for quoting, or you can always use the menu to do so. For example your editor box will look like:

    ```

>     > How big is your gpu cluster

Reviewer 2: ```suggestion
    > How big is your gpu cluster?
```

>     Our cluster is made of 256 gpus

Reviewer 2: ```suggestion
    Our cluster is made of 256 gpus.
```

    ```
    
    If you are addressing multiple comments, quote the relevant parts of each before your answer. Some people use the same comment to do multiple replies, others separate them into separate comments. Either way works. The latter approach helps for linking to a specific comment.

In general the best way to figure out what works the best is learn from Issues posted by other people - see which Issues get great responses and which get little to no response - observe what the posters who received great responses did differently from those who did not. 

> Thank you for reading this somewhat lengthy document. We would like to conclude that these are not rules, but guidelines that will help maximize the chances for us to understand what you are trying to communicate, reproduce the problem and resolve it to your satisfaction and the benefit of the whole community.

Reviewer 2: ```suggestion
Thank you for reading this somewhat lengthy document. We would like to conclude that these are not absolute rules, but friendly advice that will help maximize the chances for us to understand what you are trying to communicate, reproduce the problem then resolve it to your satisfaction and the benefit of the whole community.
```

> If after reading this document there are remaining questions on how and why or there is a need for further elucidation, please, don't hesitate to ask "How do I?" type of questions in the following sub-section of the forums: [XXX: @sgugger].

Reviewer 2: When we have a finished and ready to merge version of this, I will copy-paste it on the forums and we can link the topic here (so that will be a topic instead of a subsection).

<!---
Copyright 2020 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

# How To Request Support

> This is an Open Source Project so please be mindful that like in any other project of this kind there is no obligation to answer all requests for help.

Reviewer 3: Maybe add a second intro sentence that doesn't scare people away to ask for help. Like

```
However, we want to encourage you to ask for help whenever you think it's needed! We are happy about every question we get because it allows us to better understand your needs, possible misunderstandings, and most importantly a way for you to help make this library better. That being said, this document's main purpose ...
```

Think a sentence like this to not scare away issues would be nice. Not sure what you think @LysandreJik @sgugger @stas00

Reviewer: Yes, love your take on this!

> This document's main purpose is to provide guidelines at how you can formulate your requests to increase your chances to be understood and to get supported.

Reviewer 3: ```suggestion
This document's main purpose is to provide guidelines at how you can formulate your requests to increase your chances to be understood and to get support.
```

I might be very wrong here, but isn't it "`get support`"?

Author: either way works. I merged your suggestion manually, thank you!

   Pasting a 100-1000 lines of log into the issue is an immediate turn off, since it will take a lot of time to figure out where the pertinent parts of the log are.

   Attaching a full log can be helpful if it's done as an attachment, if it's enclosed in the following html code in the comment editor window:

   ```
   <details>
   <summary>Full log</summary>
   <pre>

   many
   lines
   go
   here

   </pre>
   </details>
   ```

   which would result in the following entry, which can be opened if desired, but otherwise takes little space.

   <details>
   <summary>Full log</summary>
   <pre>
   many
   lines
   go
   here
   </pre>
   </details>

    You could also provide a link to a pastebin service, but this is less beneficial since those links tend to expire quickly and future readers of your issue might not be able to access that log file anymore and may lack some context.

6. If this is an issue in your code, do try to reduce that code to a minimal example that still demonstrates the problem. Please ask at the forums if you have a hard time figuring how to do that. Please realize that we don't have the luxury of having time to try and understand all of your custom code.

>    If you really tried to make a short reproducible code but couldn't figure it out, it might be that having a traceback wil give the developer enough information to know what's going on. But if it is not enough and we can't reproduce the problem, we can't really solve it.

Reviewer: ```suggestion
   If you really tried to make a short reproducible code but couldn't figure it out, it might be that having a traceback will give the developer enough information to know what's going on. But if it is not enough and we can't reproduce the problem, we can't really solve it.
```
